File: Euler/EasyProblems/largest_prime_factor.cpp
#include <iostream>
#include <boost/lexical_cast.hpp>
#include <string>
#include <math.h>
#include <stdlib.h>

bool
isPrime(unsigned long number)
{
	if (number <=1)
		return false;
	
	unsigned long sqr_root = sqrt(number) + 1;
	
	for (unsigned long i = 2; i < sqr_root; ++i)
	{
		if  (!(number % i))
		{
			return false;
		}
	}
	
	return true;
}

unsigned long
largest_prime_factor(unsigned long number)
{
	//prime factor = prime number which divides an integer exactly
	unsigned long limit = number/2;
	
	unsigned long idx = 2;
	
	ldiv_t result = {0};
	
	while (idx <= limit)
	{
		//std::cout<<"idx: "<<idx<<"\n";
		result = ldiv(number, idx);
		if (!result.rem && isPrime(result.quot))
		{
			return result.quot;
		}
		++idx;
	}
	
	return 0;
}


int main(int argc, char* argv[])
{
	if (argc != 2)
	{
		std::cerr<<"Usage: ./lrg_pf <number>\n";
		return -1;
	}
	
	unsigned long number = boost::lexical_cast<unsigned long>(std::string(argv[1]));
	std::cout<<"Largest prime factor: "<<largest_prime_factor(number)<<"\n";
	
	return 0;
}

File: Euler/EasyProblems/largest_prime_factor.py
#!/usr/bin/env python

"""
The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143 ?
"""

import math
import sys

def isprime(n):
    """returns True if n is prime, False otherwise"""

    # 1 is not a prime number
    if n<=1:
        return False

    sqr_root =int(math.sqrt(n))+1
    #print "sqr_root: ",sqr_root

    for i in range(2, sqr_root):
        if n % i == 0:
            return False

    return True



def largest_prime_factor(number):
    """Largest prime factor
    prime factor = prime number which divides an integer exactly
    """
    limit = int(number/2)
    idx = 2
    while idx <= limit:
        a, b = divmod(number, idx)
        #if division was exact and quotient is prime
        if not b and isprime(a):
            return a
        idx += 1
#        print idx


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print "Usage: ./largest_prime_factor.py <number>"
        sys.exit(1)
    number = int(sys.argv[1])

    print "largest prime factor: {0}".format(largest_prime_factor(number))

